Fixes #36495 - Extend Windows templates for Puppet and Ansible

nadjaheitmann opened the pull request:
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

Commit:
Fixes #36495 - Extend Windows templates for Puppet and Ansible

Co-authored-by: Fabrice Brimioulle <brimioulle@atix.de>
Co-authored-by: Ewoud Kohl van Wijngaarden <ewoud@kohlvanwijngaarden.nl>

sbernhard commented on app/views/unattended/provisioning_templates/script/windows_default_script.erb:
```diff
@@ -21,13 +24,25 @@ set WGET=wget64.exe
+echo Setting network config for static interfaces
+    netsh interface ip set address name="Ethernet0" static "<%= iface.ip %>" "<%= iface.subnet.mask %>" "<%= iface.subnet.gateway %>"
+    net stop "DHCP-Client"
+    net stop "DHCP Client"
```
why 2 times?

jtruestedt replied:
Because the service has different names depending on Windows-Version (i am not sure if it just the major version or also language-dependency).
Same applies for DNS

jtruestedt commented on app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb:
```diff
@@ -12,13 +12,15 @@ description: |
   system_locale = host_param('systemLocale') ? host_param('systemLocale') : 'en-US'
   system_ui_lang = host_param('systemUILanguage') ? host_param('systemUILanguage') : 'en-US'
   setup_ui_lang = 'en-US'
+  input_locale = host_param('inputLocale') ? host_param('inputLocale'): 'de-DE'
```
Do you really want to set the german keyboard-layout as the default?
I would use the en-US one as default for the foreman-community

nadjaheitmann replied:
True, thanks!

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -33,6 +33,7 @@ description: |
   puppet_enabled = !host_param_true?('skip-puppet-setup') && (host_puppet_server.present? || host_param_true?('force-puppet'))
   salt_enabled = host_param('salt_master') ? true : false
   chef_enabled = @host.respond_to?(:chef_proxy) && @host.chef_proxy
+  network_location = host_param('networklocation') ? host_param('networklocation') : 'Private'
```
You can provide a default to `host_param`
```suggestion
  network_location = host_param('networklocation', 'Private')
```

This is repeated a few times. I'm only commenting once, but please fix it in the other files too.

nadjaheitmann replied:
@ekohl Thanks for reviewing! I was just looking for `host_param` in the templates and I did not find any template that makes use of the default. They are using either the notation above or something like 
```suggestion
  network_location = host_param('networklocation') || 'Private'
```
Readability wise I like that the most, although it kind of makes sense to use the default value if exists. Any thoughts?

ekohl replied:
Internally it already calls `host.host_param(param_name) || default` so your latter suggestion can also work. I'm still leaning to `host_param('networklocation', 'Private')` to match `.fetch()` but we don't implement block support so I can see arguments for both.

nadjaheitmann replied:
I have implemented your suggestion. I feel that it looks a bit strange. 

Finally, I think it would be a good idea to unify the notation for all the templates.

nadjaheitmann replied:
@ekohl @sbernhard You outvoted me, we'll go with @ekohl 's suggestion :) I hope I did not miss any occurences. Please have a look if anything else needs changing. I suggest to open an issue for unifying the usage of `host_param` and `host_param_true?` within the unattended templates. Any objections?

sbernhard commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -31,8 +31,9 @@ description: |
   # safemode renderer does not support unary negation
   puppet_enabled = !host_param_true?('skip-puppet-setup') && (host_puppet_server.present? || host_param_true?('force-puppet'))
-  salt_enabled = host_param('salt_master') ? true : false
+  salt_enabled = !host_param('salt_master').nil?
```
hmm, " " would be 'not nil', too. Maybe use this: !host_param('salt_master').to_s.empty?

ekohl replied:
I think you should treat empty strings as not enabled either
```suggestion
  salt_enabled = host_param('salt_master').present?
```

nadjaheitmann commented:
@ekohl Anything else that should be improved?

nadjaheitmann commented:
@ekohl @sbernhard Can we move forward here?

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -44,89 +45,104 @@ description: |
 
-  <% if host_param('ansible_winrm_transport') == 'basic' %>
+  <% if host_param('ansible_user').present? %>
+    <% if host_param('create_ansible_user') %>
```
If you use `host_param_true?` you can set it to `true` on the host group level and `false` on the host level.
```suggestion
    <% if host_param_true?('create_ansible_user') %>
```

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -44,89 +45,104 @@ description: |
+    echo Configuring Puppet
+    mkdir "C:\ProgramData\PuppetLabs\puppet\etc"
+    echo [main] > "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
+    echo server=http://<%= @host.puppetmaster %> >> "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
```
I think this shouldn't be needed (see below), but if it is: server is the hostname, there's no protocol. Also, `puppet_puppet_server` is the current macro. `@host.puppetmaster is deprecated.
```suggestion
    echo server=<%= host_puppet_server %> >> "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
```

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -44,89 +45,104 @@ description: |
+    echo Downloading Puppet installer
+    wget "<%= host_param('win_puppet_source') %>" -O C:\puppet-agent-x64-latest.msi
+    echo Configuring Puppet
+    mkdir "C:\ProgramData\PuppetLabs\puppet\etc"
+    echo [main] > "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
```
Why is this needed now? I suspected the `puppet.msi` (which is still there) would be sufficient.

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -44,89 +45,104 @@ description: |
+    echo server=http://<%= @host.puppetmaster %> >> "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
+    echo autoflush=true >> "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
+    echo Installing Puppet
     start /w "" msiexec /qn /i C:\extras\puppet.msi PUPPET_AGENT_STARTUP_MODE=Manual PUPPET_MASTER_SERVER=<%= host_puppet_server -%> PUPPET_AGENT_ACCOUNT_DOMAIN=<%= @host.domain -%> PUPPET_AGENT_ACCOUNT_USER=administrator PUPPET_AGENT_ACCOUNT_PASSWORD="<%= host_param('domainAdminAccountPasswd') -%>"
```
I suspect `PUPPET_MASTER_SERVER` is wrong here. Reading https://www.puppet.com/docs/puppet/7/install_agents#msi_properties I think it should be:
```suggestion
    start /w "" msiexec /qn /i C:\extras\puppet.msi PUPPET_AGENT_STARTUP_MODE=Manual PUPPET_SERVER=<%= host_puppet_server -%> PUPPET_AGENT_ACCOUNT_DOMAIN=<%= @host.domain -%> PUPPET_AGENT_ACCOUNT_USER=administrator PUPPET_AGENT_ACCOUNT_PASSWORD="<%= host_param('domainAdminAccountPasswd') -%>"
```

It should also set `PUPPET_CA_SERVER=<%= host_puppet_ca_server>%`.

sbernhard replied:
Can you add this PUPPET_CA_SERVER @nadjaheitmann ? Does this make sense?

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -44,89 +45,104 @@ description: |
+    mkdir "C:\ProgramData\PuppetLabs\puppet\etc"
+    echo [main] > "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
+    echo server=http://<%= @host.puppetmaster %> >> "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
+    echo autoflush=true >> "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
```
I had to look this up, but https://www.puppet.com/docs/puppet/7/configuration.html#autoflush says this should be default:
```suggestion
```

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -44,89 +45,104 @@ description: |
   <% if puppet_enabled %>
-    echo Installing puppet
+    echo Downloading Puppet installer
+    wget "<%= host_param('win_puppet_source') %>" -O C:\puppet-agent-x64-latest.msi
```
This doesn't match the path below (`C:\extras\puppet.msi`) so likely to cause problems.

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -44,89 +45,104 @@ description: |
   <% end%>
 
-  <% if host_param('foremanDebug') != true %>
+  <% if !host_param_true('foremanDebug') %>
```
```suggestion
  <% if !host_param_true?('foremanDebug') %>
```

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -44,89 +45,104 @@ description: |
+    rd /s /q c:\drivers
+    rd /s /q c:\updates
+
+    <% if puppet_enabled %>
```
```suggestion
    <% if puppet_enabled -%>
```

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -44,89 +45,104 @@ description: |
     <% end %>
   <% end %>
 
+  <%#
+  Adjust power setting to 'High performance'
+  -%>
+  powershell /c "powercfg /setactive 8c5e7fda-e8bf-4a96-9a85-a6e23a8c635c"
```
Is this really appropriate for this PR? I can't speak for Windows admins since I rarely touch it, but doing this unconditionally may not be correct for users.

ekohl reviewed (dismissed):
Please consistently use `-%>`, at least for newly added blocks.

nadjaheitmann commented:
> Please consistently use `-%>`, at least for newly added blocks.

Thanks, I'll try to unify the templates. Working on the other changes as well, thank you for your input! Highly appreciated!

nadjaheitmann commented:
@ekohl I cleaned up the tags (hope I got them all) and modified the configurations in question. Right now, there are three commits for better visibility of the changes. Do you mind having another look?

nadjaheitmann commented:
@ekohl Anything else that needs fixing?

nadjaheitmann commented on test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Windows_default_finish.windows10_dhcp.snap.txt:
```diff
@@ -0,0 +1,79 @@
+  echo Syncing time
+  w32tm /resync
+  w32tm /resync
+
+  
+  
+  
+      powershell /c "Get-NetConnectionProfile -InterfaceAlias \"Ethernet0\" | Set-NetConnectionProfile -NetworkCategory Private"
+  
+  
+  
+  
+      echo Downloading Puppet installer
+    wget "" -O C:\puppet-agent-x64-latest.msi
+    echo Installing Puppet
+    start /w "" msiexec /qn /i C:\puppet-agent-x64-latest.msi PUPPET_AGENT_STARTUP_MODE=Manual PUPPET_SERVER= PUPPET_AGENT_ACCOUNT_DOMAIN=snap.example.com PUPPET_AGENT_ACCOUNT_USER=administrator PUPPET_AGENT_ACCOUNT_PASSWORD=""
+    echo Setting Puppet to auto start
+    sc config puppet start= auto
+    sc query puppet
+  
+      echo Rebooting in 60 sec
+    shutdown /r /t 60
+
+    echo Removing wimaging files
+    rd /s /q c:\wimaging
+    sdelete.exe -accepteula -p 2 c:\Windows\Panther\unattend.xml
+    sdelete.exe -accepteula -p 2 C:\Windows\Setup\Scripts\SetupComplete.cmd
+
+    echo Removing leftover directories
+    rd /s /q c:\MININT
+    rd /s /q c:\drivers
+    rd /s /q c:\updates
+
+            echo Removing Puppet installer
+        sdelete.exe -accepteula -p 2 C:\puppet-agent-x64-latest.msi
+    
+    echo Removing deploy directory
+    rd /s /q c:\deploy
+
+  
```
Indentation-wise, this template snapshot looks kind of funny....

nadjaheitmann replied:
fixed indentation

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -31,102 +33,108 @@ description: |
+<% end -%>
+
+echo Syncing time
+w32tm /resync
+w32tm /resync
```
Is it intentional that you call this twice?

nadjaheitmann replied:
I removed the second call.

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -31,102 +33,108 @@ description: |
+<% if host_param('computerDomain') -%>
+<% if host_param('domainAdminAccount').present? && host_param('domainAdminAccountPasswd').present? -%>
+echo Performing secure domain join
+powershell.exe -OutputFormat text -command Add-Computer -DomainName '<%= host_param('computerDomain') -%>' -Credential (New-Object -TypeName System.Management.Automation.PSCredential '<%= host_param('domainAdminAccount') -%>', (ConvertTo-SecureString -String '<%= host_param('domainAdminAccountPasswd') -%>' -AsPlainText -Force)) <% if host_param('computerOU').present? -%>-OUPath '<%= host_param('c ...
```
Nit: perhaps avoid trailing whitespace?
```suggestion
powershell.exe -OutputFormat text -command Add-Computer -DomainName '<%= host_param('computerDomain') -%>' -Credential (New-Object -TypeName System.Management.Automation.PSCredential '<%= host_param('domainAdminAccount') -%>', (ConvertTo-SecureString -String '<%= host_param('domainAdminAccountPasswd') -%>' -AsPlainText -Force))<% if host_param('computerOU').present? -%> -OUPath '<%= host_param('computerOU') -%>'<% end -%>
```

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -31,102 +33,108 @@ description: |
+
+<% if puppet_enabled -%>
+echo Downloading Puppet installer
+wget "<%= host_param('win_puppet_source') %>" -O C:\puppet-agent-x64-latest.msi
+echo Installing Puppet
+start /w "" msiexec /qn /i C:\puppet-agent-x64-latest.msi PUPPET_AGENT_STARTUP_MODE=Manual PUPPET_SERVER=<%= host_puppet_server -%> PUPPET_CA_SERVER=<%= host_puppet_ca_server -%> PUPPET_AGENT_ACCOUNT_DOMAIN=<%= @host.domain -%> PUPPET_AGENT_ACCOUNT_USER=administrator PUPPET_AGENT_ACCOUNT_PASSWORD="<%= host_param('domainAdminAccountPasswd') -%>"
```
With the eye on non-x64 support (like ARM), would this be better?
```suggestion
wget "<%= host_param('win_puppet_source') %>" -O C:\puppet-agent-latest.msi
echo Installing Puppet
start /w "" msiexec /qn /i C:\puppet-agent-latest.msi PUPPET_AGENT_STARTUP_MODE=Manual PUPPET_SERVER=<%= host_puppet_server -%> PUPPET_CA_SERVER=<%= host_puppet_ca_server -%> PUPPET_AGENT_ACCOUNT_DOMAIN=<%= @host.domain -%> PUPPET_AGENT_ACCOUNT_USER=administrator PUPPET_AGENT_ACCOUNT_PASSWORD="<%= host_param('domainAdminAccountPasswd') -%>"
```

ekohl commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -31,102 +33,108 @@ description: |
+
+<% if puppet_enabled -%>
+echo Removing Puppet installer
+sdelete.exe -accepteula -p 2 C:\puppet-agent-x64-latest.msi
```
And if you change the filename above, then here too
```suggestion
sdelete.exe -accepteula -p 2 C:\puppet-agent-latest.msi
```

nadjaheitmann replied:
Note: If I change the name here, I also need to change it in the windows user data template.

ekohl commented on app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb:
```diff
@@ -70,9 +74,23 @@ description: |
                     <Value><%= root_pass %></Value>
+                    <PlainText>false</PlainText>
                 </AdministratorPassword>
+            <% if host_param('ansible_user') and host_param('create_ansible_user') -%>
```
In Ruby `&&` is preferred over `and`
```suggestion
            <% if host_param('ansible_user') && host_param('create_ansible_user') -%>
```

ekohl commented on app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb:
```diff
@@ -70,9 +74,23 @@ description: |
+                <LocalAccounts>
+                    <LocalAccount wcm:action="add">
+                        <Password>
+                            <Value><%= host_param('ansible_ssh_pass') %></Value>
```
I wonder if anyone has `<` in their password and you need to XML escape values

nadjaheitmann replied:
Yeah, you are right, we need to fix this as well.

nadjaheitmann replied:
Added HTML escaping for Windows https://www.w3.org/TR/REC-xml/#sec-cdata-sect

ekohl commented on app/views/unattended/provisioning_templates/script/windows_default_script.erb:
```diff
@@ -4,15 +4,20 @@ name: Windows peSetup.cmd
 <%
-  proxy_uri = host_param('http-proxy') ? "http://#{host_param('http-proxy')}:#{host_param('http-proxy-port') || 3128}" : nil
+  iface = @host.provision_interface
+  proxy_uri = host_param('http-proxy') ? "#{host_param('http-proxy')}:#{host_param('http-proxy-port')}" : nil
```
Are you intentionally dropping the default value?
```suggestion
  proxy_uri = host_param('http-proxy') ? "#{host_param('http-proxy')}:#{host_param('http-proxy-port', 3128)}" : nil
```

ekohl commented on test/unit/foreman/renderer/snapshots/ProvisioningTemplate/script/Windows_peSetup.cmd.windows10_dhcp.snap.txt:
```diff
@@ -0,0 +1,97 @@
+@echo off
+set WGET=wget64.exe
+
+zerombr
```
Looks like it's using a Red Hat partitioning table. That's probably a bug in the factory.

nadjaheitmann replied:
I have added a partition table for windows. I am absolutely not sure if I added a correct format, though.

ekohl commented on app/services/foreman/template_snapshot_service.rb:
```diff
@@ -205,6 +209,14 @@ def rocky9_dhcp
     end
 
+    def windows10_dhcp
+      host = FactoryBot.build(:host_for_snapshots_ipv4_dhcp_windows10,
```
As noted earlier: this ends up with a Red Hat partition table. Perhaps this is also a problem for the Debian factories.

nadjaheitmann replied:
I found that there is a partition table configured for Debian systems but I have not checked how thoroughly it is used. https://github.com/theforeman/foreman/blob/develop/test/factories/host_related.rb#L23

ekohl reviewed (commented):
Overall I think this looks pretty good and many notes are more nit picks.

nadjaheitmann commented on test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Windows_default_finish.windows10_dhcp.snap.txt:
```diff
@@ -0,0 +1,78 @@
+echo Downloading Puppet installer
+wget "" -O C:\puppet-agent-x64-latest.msi
+echo Installing Puppet
+start /w "" msiexec /qn /i C:\puppet-agent-x64-latest.msi PUPPET_AGENT_STARTUP_MODE=Manual PUPPET_SERVER= PUPPET_CA_SERVER= PUPPET_AGENT_ACCOUNT_DOMAIN=snap.example.com PUPPET_AGENT_ACCOUNT_USER=administrator PUPPET_AGENT_ACCOUNT_PASSWORD=""
```
What looks funny to me is that `PUPPET_SERVER=` and `PUPPET_CA_SERVER=` are set without any values here. @ekohl Is this okay for the test or is it possible to set default values for host_puppet_server and host_puppet_ca_server?

nadjaheitmann commented:
> Overall I think this looks pretty good and many notes are more nit picks.

Thanks again @ekohl for the review. I addressed all the issues you mentioned. Do you mind having another look?

nadjaheitmann commented:
@ekohl More changes required? :)

stejskalleos commented:
Thanks @nadjaheitmann !
